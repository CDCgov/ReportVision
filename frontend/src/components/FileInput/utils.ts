export const makeSafeForID = (name: string): string => {
    return name.replace(/[^a-z0-9]/g, function replaceName(s) {
      const c = s.charCodeAt(0)
      if (c === 32) return '-'
      if (c >= 65 && c <= 90) return `img_${s.toLowerCase()}`
      return `__${c.toString(16).slice(-4)}`
    })
  }