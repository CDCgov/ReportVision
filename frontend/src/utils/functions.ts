import { CustomShape } from "../contexts/AnnotationContext";
import { ImageData } from "../pages/AnnotateTemplate";
// import {useCreateTemplateStore} from "../types/templates.ts";

export const makeScreenshots = async (images, shapes) => {
    // const images: ImageData[] = JSON.parse(localStorage.getItem('images') || '[]') as ImageData[];
    // const shapes: CustomShape[][] = JSON.parse(localStorage.getItem('shapes') || '[]') as CustomShape[][];
    const screenshots: string[] = [];

    for (let i = 0; i < images.length; i++) {
        try {
            const screenshot = await createScreenshot(images[i], shapes[i] ?? []);
            screenshots.push(screenshot);
        } catch (e) {
            console.log(e);
        }
    }

    // Final log of all screenshots
    // localStorage.setItem('screenshots', JSON.stringify(screenshots));
    return screenshots;
};

const createScreenshot = (localImg: ImageData, shapeData: CustomShape[]) => {
    return new Promise<string>((resolve) => {
        const canvas = document.createElement('canvas');
        const context = canvas.getContext('2d');
        if (context) {
            canvas.width = Number(localImg.width);
            canvas.height = Number(localImg.height);
            const img = new Image();
            img.src = localImg.image;

            img.onload = () => {
                context.drawImage(img, 0, 0, canvas.width, canvas.height);
                shapeData.forEach((shape) => {
                    context.fillStyle = shape.color || 'red';

                    const xCoords = shape.points.map(point => point[0]);
                    const yCoords = shape.points.map(point => point[1]);

                    const xMin = Math.min(...xCoords);
                    const yMin = Math.min(...yCoords);
                    const xMax = Math.max(...xCoords);
                    const yMax = Math.max(...yCoords);

                    const width = xMax - xMin;
                    const height = yMax - yMin;

                    // Draw the rectangle based on the calculated coordinates
                    context.fillRect(xMin, yMin, width, height);
                });

                const screenshot = canvas.toDataURL('image/png');
                resolve(screenshot);
                canvas.remove();
                context.closePath();
            };
            // If the image fails to load, reject the promise
            img.onerror = () => {
                console.error('Failed to load image');
                resolve('');
                context.closePath();
                canvas.remove();
            }
        }
    });
};