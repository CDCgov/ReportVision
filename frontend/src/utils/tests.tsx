import { ReactNode } from "react";
import { BrowserRouter } from "react-router-dom";
import { FilesProvider } from "../contexts/FilesContext";
import {QueryClient, QueryClientProvider} from "@tanstack/react-query";

const queryClient = new QueryClient();

export const Wrapper = ({ children }: { children: ReactNode }) => {
    return <QueryClientProvider client={queryClient}><BrowserRouter><FilesProvider>{children}</FilesProvider></BrowserRouter></QueryClientProvider>;
}