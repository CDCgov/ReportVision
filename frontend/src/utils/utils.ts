export function base64ToBlob(base64Data: string): Blob {
  // Remove the data URL part if it exists
  const base64String = base64Data.split(",")[1];

  // Decode base64 string into binary data
  const binaryString = atob(base64String);

  // Create a Uint8Array from the binary string
  const binaryData = new Uint8Array(binaryString.length);
  for (let i = 0; i < binaryString.length; i++) {
    binaryData[i] = binaryString.charCodeAt(i);
  }

  // Create a Blob from the binary data
  const blob = new Blob([binaryData], { type: "image/png" });

  return blob;
}

export function convertBase64SvgToBase64Png(
  base64Svg: string,
  width: number,
  height: number,
): Promise<string> {
  return new Promise((resolve, reject) => {
    // Create an Image object
    const img: HTMLImageElement = new Image();

    // Set the image source to the base64-encoded SVG
    img.src = `data:image/svg+xml;base64,${base64Svg}`;

    img.onload = () => {
      // Create a canvas element
      const canvas: HTMLCanvasElement = document.createElement("canvas");
      canvas.width = width;
      canvas.height = height;
      const ctx: CanvasRenderingContext2D | null = canvas.getContext("2d");

      if (!ctx) {
        reject(new Error("Failed to get canvas context"));
        return;
      }

      // Draw the SVG image onto the canvas
      ctx.drawImage(img, 0, 0, width, height);

      // Convert the canvas content to a PNG base64 data URL
      const base64Png = canvas.toDataURL("image/png");

      // Remove the prefix ("data:image/png;base64,") and return the base64 string
      resolve(base64Png.replace(/^data:image\/png;base64,/, ""));
    };

    img.onerror = () => {
      reject(new Error("Failed to load SVG image"));
    };
  });
}

export const convertPdfToImages = async (
  file: File,
  pdfjsLib: any,
): Promise<string[]> => {
  const images: Array<string> = [];
  const data = URL.createObjectURL(file);

  const pdf = await pdfjsLib.getDocument(data).promise;
  const canvas = document.createElement("canvas");
  for (let i = 0; i < pdf.numPages; i++) {
    const page = await pdf.getPage(i + 1);
    const viewport = page.getViewport({ scale: 1 });
    const context = canvas.getContext("2d")!;
    canvas.height = viewport.height;
    canvas.width = viewport.width;
    await page.render({ canvasContext: context, viewport: viewport }).promise;
    images.push(canvas.toDataURL());
  }
  canvas.remove();
  URL.revokeObjectURL(data);
  return images;
};
