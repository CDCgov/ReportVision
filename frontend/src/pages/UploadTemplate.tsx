import { useNavigate } from "react-router-dom"
import { Divider } from "../components/Divider"
import { UploadHeader } from "../components/Header"
import { Stepper } from "../components/Stepper"
import { Uploadfile } from "../components/UploadFile"
import { AnnotateStep } from "../utils/constants"
import { useFiles } from "../contexts/FilesContext"

export const UploadTemplate = () => {
    const navigate = useNavigate();
    const { files } = useFiles();
    return (
        <div style={{ display: 'flex', flexDirection: 'column', justifyContent: 'start', width: '100%', height: '100%', padding: '16px 8px, 8px, 8px' }}>
            <UploadHeader title="Annotate new template" onBack={() => {navigate('/')}} onSubmit={() => navigate('/new-template/annotate')} isUpload={files.length === 0} />
            <Divider margin="0px" />
            <div style={{ display: 'flex', justifyContent: 'center', paddingTop: '32px'}}>
                <Stepper currentStep={AnnotateStep.Upload} />
            </div>
            <Divider margin="0px" />
            <Uploadfile />
        </div>
    )
}