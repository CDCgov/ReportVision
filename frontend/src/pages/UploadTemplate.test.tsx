import { render } from "@testing-library/react";
import { describe, it, expect } from "vitest";
import { UploadTemplate } from "./UploadTemplate";
import { Wrapper } from "../utils/tests";

describe("UploadTemplate component", () => {
  it("matches the snapshot", () => {
    const { asFragment } = render(<UploadTemplate />, { wrapper: Wrapper });
    expect(asFragment()).toMatchSnapshot();
  });
});
