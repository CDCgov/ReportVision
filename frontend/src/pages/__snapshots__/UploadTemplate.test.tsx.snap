// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UploadTemplate component > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    style="display: flex; flex-direction: column; justify-content: start; width: 100%; height: 100%;"
  >
    <header
      class="usa-header header"
      data-testid="header"
    >
      <div
        class="display-flex height-full flex-align-center flex-justify"
      >
        <div
          class="display-flex flex-jusitfy-start height-full flex-align-center flex-justify"
        >
          <button
            class="usa-button usa-button--unstyled close-button"
            data-testid="close-button"
            type="button"
          >
            <svg
              class="usa-icon usa-icon--size-3"
              color="black"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </button>
          <h1>
            Annotate new template
          </h1>
        </div>
        <div
          class="display-flex flex-jusitfy-start height-full flex-align-center"
        >
          <button
            class="usa-button usa-button--outline reset-button"
            data-testid="button"
            type="reset"
          >
            Back
          </button>
          <button
            class="usa-button usa-button--base"
            data-testid="button"
            disabled=""
            type="submit"
          >
            Submit
          </button>
        </div>
      </div>
    </header>
    <div
      style="border-bottom: 1px solid #ccc; margin: 0px;"
    />
    <div
      style="display: flex; justify-content: center; padding-top: 32px;"
    >
      <div
        aria-label="progress"
        class="usa-step-indicator"
        data-testid="step-indicator"
      >
        <ol
          class="usa-step-indicator__segments"
        >
          <li
            aria-current="true"
            class="usa-step-indicator__segment usa-step-indicator__segment--current"
          >
            <span
              class="usa-step-indicator__segment-label"
            >
              Upload New Segment 
            </span>
          </li>
          <li
            class="usa-step-indicator__segment"
          >
            <span
              class="usa-step-indicator__segment-label"
            >
              Annotate 
              <span
                class="usa-sr-only"
              >
                not completed
              </span>
            </span>
          </li>
          <li
            class="usa-step-indicator__segment"
          >
            <span
              class="usa-step-indicator__segment-label"
            >
              Save Template 
              <span
                class="usa-sr-only"
              >
                not completed
              </span>
            </span>
          </li>
        </ol>
        <div
          class="usa-step-indicator__header"
        >
          <h5
            class="usa-step-indicator__heading"
            style="display: none;"
          >
            <span
              class="usa-step-indicator__heading-counter"
            >
              <span
                class="usa-sr-only"
                data-testid="step-text"
              >
                Step
              </span>
              <span
                class="usa-step-indicator__current-step"
              >
                1
              </span>
               
              <span
                class="usa-step-indicator__total-steps"
              >
                of 3
              </span>
               
            </span>
            <span
              class="usa-step-indicator__heading-text"
            >
              Upload New Segment
            </span>
          </h5>
        </div>
      </div>
    </div>
    <div
      style="border-bottom: 1px solid #ccc; margin: 0px;"
    />
    <div
      class="display-flex flex-column flex-align-center flex-justify-start height-full width-full padding-2 bg-primary-lighter"
    >
      <div
        class="upload-file-header-content-container"
      >
        <h1
          class="upload-file-header"
        >
          Upload new file to annotate
        </h1>
        <p>
          You can upload new lab files (PDF s or images) to annotate and save as a template. Annotation is the process of marking and labeling specific areas of an image to identify key elements or features.
        </p>
        <p
          class="upload-file-instructions"
        >
          Select one file
        </p>
        <div
          class="width-full height-full bg-white dashed-container dashed-container-preupload"
          data-testid="dashed-container"
        >
          <div
            class="display-flex flex-column flex-align-center width-full height-full"
          >
            <div
              class="usa-file-input padding-bottom-2 width-full height-full"
              data-testid="file-input"
            >
              <div
                class="usa-file-input__target"
                data-testid="file-input-droptarget"
              >
                <div
                  aria-hidden="true"
                  class="usa-file-input__instructions"
                  data-testid="file-input-instructions"
                >
                  <span
                    class="usa-file-input__drag-text"
                  >
                    Drag files here or
                  </span>
                  <span
                    class="usa-file-input__choose"
                  >
                     choose from folder
                  </span>
                </div>
                <div
                  class="usa-file-input__box"
                  data-testid="file-input-box"
                />
                <input
                  class="usa-file-input__input"
                  data-testid="file-input-input"
                  id="file-input-:r0:"
                  name="file-input-single"
                  type="file"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="display-flex flex-jusitfy-start flex-align-center"
        >
          <button
            class="usa-button usa-button--outline upload-back-button"
            data-testid="button"
            type="reset"
          >
            Cancel Import
          </button>
          <button
            class="usa-button upload-annotate-button"
            data-testid="button"
            disabled=""
            type="submit"
          >
            Annotate
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
